src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -712,28 +706,38 @@ private TypeSymbol ErrorType(string name = "")
         /// decision when no decision appears to match. This implementation is nonrecursive to avoid
         /// overflowing the compiler's evaluation stack when compiling a large switch statement.
         /// </summary>
-        private BoundDecisionDag MakeBoundDecisionDag(SyntaxNode syntax, ArrayBuilder<StateForCase> cases)
+        private BoundDecisionDag MakeBoundDecisionDag(SyntaxNode syntax, ref TemporaryArray<StateForCase> cases)
         {
-            // A work list of DagStates whose successors need to be computed
-            var workList = ArrayBuilder<DagState>.GetInstance();
+            // A mapping used to make each DagState unique (i.e. to de-dup identical states).
+            PooledDictionary<DagState, DagState> uniqueState = s_uniqueStatePool.Allocate();
 
             // Build the state machine underlying the decision dag
-            DecisionDag decisionDag = MakeDecisionDag(cases, workList);
+            var decisionDag = MakeDecisionDag(ref cases, uniqueState);
 
-            workList.Free();
+            // Compute the bound decision dag corresponding to each node of decisionDag, and store
+            // it in node.Dag.
+            var defaultDecision = new BoundLeafDecisionDagNode(syntax, _defaultLabel);
```

Comment:
nit: the order was changed here, but that doesn't seem necessary

Reply:
Restored.